### PR TITLE
Editor: enhance Location Selector for more granular selection (7459)

### PR DESCRIPTION
**Pull request by popthestack**
This adds an embedded Google Map to the post editor sidebar to allow for more granular location selection. Clicking the map sets the geolocation for the post.

This is for issue #7459 and still a work in progress. Opening this PR for feedback.

One thing I'm not sure about is how to handle the API key. Anyone running it locally would have to use their own, as putting an API key directly in the source is a bad idea. I'm thinking the build process for the official Calypso release would need to include an Automattic API key, but I don't know how that would be handled.

Here's a gif of this in action:

google-map-enhanced-location

### CODE REVIEW

File: client/post-editor/editor-location/map.jsx [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+
+	componentDidMount() {
+		// Connect the initMap() function within this class to the global window
+		// context, so Google Maps can invoke it.
+		window.initMap = this.initMap.bind( this );
+
+		const src =
+			'https://maps.googleapis.com/maps/api/js?' +
+			qs.stringify( {
+				key: 'INSERT-API-KEY-HERE',
```
Comment (timmyc) [inline client/post-editor/editor-location/map.jsx]: Does the loading of the request fail without the `key` param?

File: client/post-editor/editor-location/map.jsx [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+						lat: event.latLng.lat(),
+						lng: event.latLng.lng(),
+					},
+				},
+			} );
+		} );
+	}
+
+	componentDidUpdate() {
+		if ( this.map && this.props.coordinates ) {
```
Comment (timmyc) [inline client/post-editor/editor-location/map.jsx]: could `this.props.coordinates` ever be an empty array here? Also should we consider deconstructing `const { coordinates } = this.props;` to clean things up a a bit?

File: client/post-editor/editor-location/map.jsx [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+		if ( this.marker ) {
+			this.marker.setMap( null );
+			this.marker = undefined;
+		}
+	}
+
+	render() {
+		const mapRef = el => ( this.mapRef = el );
+		return (
+			<div>
```
Comment (timmyc) [inline client/post-editor/editor-location/map.jsx]: Is the wrapping div needed here?

File: client/post-editor/editor-location/map.jsx (hunk 1 above)
Comment (popthestack) [reply client/post-editor/editor-location/map.jsx]: Yeah, it looks like this:

screenshot 2017-11-27 22 43 52

I'm thinking it'd be good to fall back on the old map if there's no API key. I'll go ahead and implement that.

File: client/post-editor/editor-location/map.jsx (hunk 3 above)
Comment (popthestack) [reply client/post-editor/editor-location/map.jsx]: No, but it looked like the standard way of rendering components in Calypso, even for simple stuff. I can change it though.

File: client/post-editor/editor-location/map.jsx (hunk 2 above)
Comment (popthestack) [reply client/post-editor/editor-location/map.jsx]: Good point. The `EditorLocation` component won't render the map if coordinates is falsy so a falsy check here is pointless. So far I've only seen `this.map` undefined, but it'd best to check the length of the coordinates.

File: client/post-editor/editor-location/map.jsx (hunk 3 above)
Comment (timmyc) [reply client/post-editor/editor-location/map.jsx]: Yeah in this instance here I don't think it is needed. Unless we have a className on there for styling purposes - though the map dimensions are hard coded.

**Comment by kwight** [comment]
> One thing I'm not sure about is how to handle the API key... I'm thinking the build process for the official Calypso release would need to include an Automattic API key, but I don't know how that would be handled.

Hm, interesting; I can't think of an existing case in Calypso in which we would have to handle an individual user's API key. On dotcom, we've got the Contact Info widget which allows a form field for entering a user's own API key (background at p5j4vm-1gT-p2). Generally, asking a user to provide an API key is an awful experience on their part. We _could_ just add a form field to Settings > General... @timmyc any ideas on the API key handling?

**Comment by timmyc** [comment]
In the context of this PR, we shouldn't need an API key since the domain has been "grandfathered" in so-to-speak. For example, we use the gmaps api on the stats page to render the country views widget without a key.

So I _think_ we should be okay here too

**Comment by dmsnell** [comment]
What's the status of this PR? Is there something anyone is waiting on in order to move it forward or to continue to work on it? Thanks for all the work! The location selector is important to me and it's needed considerable love for a long time. :+1:

**Comment by popthestack** [comment]
I haven't forgotten about this! Life got crazy with the holidays. I'm hoping to put some time into wrapping it up later this month.

**Comment by dmsnell** [comment]
Also, is Google Maps the best choice here? Would he have access to something like OpenStreetMaps with similar effort? Would that be a positive or negative shift in experience?

**Comment by simison** [comment]
Just to throw in some food for thought —

>Also, is Google Maps the best choice here? Would he have access to something like OpenStreetMaps with similar effort? Would that be a positive or negative shift in experience?

I believe in our scale [MapBox](https://www.mapbox.com/) would be way cheaper than Google Maps and their [ethos](https://www.mapbox.com/about/open/) are much closer to Automattic's. 

Tho it seems like Calypso [is already using](https://github.com/Automattic/wp-calypso/search?utf8=%E2%9C%93&q=maps&type=) Google Maps all over the place so the move would need to be much more coordinated.

**Comment by popthestack** [comment]
@alisterscott Yes, I would like some help. I don't have a good idea of when I'll get back to this. I was hoping to have tackled it by now. Any help anyone wants to contribute would be awesome.

**Comment by vindl** [comment]
I'm going to close out this one since it looks like no one is actively working on it anymore. If that's not the case or if someone would like to continue working on it in the future, feel free to reopen it.

**Pull request closed without merging**